ql/methods/finitedifferences/operators/fdmcirop.cpp

```diff
@@ -0,0 +1,183 @@
+*/
+
+#include <ql/methods/finitedifferences/meshers/fdmmesher.hpp>
+#include <ql/methods/finitedifferences/operators/fdmcirop.hpp>
+#include <ql/methods/finitedifferences/operators/fdmlinearoplayout.hpp>
+#include <ql/methods/finitedifferences/operators/secondderivativeop.hpp>
+#include <ql/methods/finitedifferences/operators/secondordermixedderivativeop.hpp>
+#include <ql/processes/blackscholesprocess.hpp>
+
+namespace QuantLib {
+
+    FdmCIREquityPart::FdmCIREquityPart(
+        const ext::shared_ptr<FdmMesher>& mesher,
+        const ext::shared_ptr<GeneralizedBlackScholesProcess>& bsProcess,
+        Real strike)
+    : dxMap_ (FirstDerivativeOp(0, mesher)),
+      dxxMap_(SecondDerivativeOp(0, mesher)),
+      mapT_ (0, mesher),
+      mesher_(mesher),
+      qTS_(bsProcess->dividendYield().currentLink()),
+      strike_(strike),
+      sigma1_(bsProcess->blackVolatility().currentLink()){
+    }
+
+    void FdmCIREquityPart::setTime(Time t1, Time t2) {
+        const Rate q = qTS_->forwardRate(t1, t2, Continuous).rate();
+
+        const Real v = sigma1_->blackForwardVariance(t1, t2, strike_)/(t2-t1);
+
+        mapT_.axpyb(mesher_->locations(1) - q - 0.5*v, dxMap_,
+                        dxxMap_.mult(Array(mesher_->layout()->size(), v/2)), -0.5*mesher_->locations(1));
+    }
+
+    const TripleBandLinearOp& FdmCIREquityPart::getMap() const {
+        return mapT_;
+    }
+
+    FdmCIRRatesPart::FdmCIRRatesPart(
+        const ext::shared_ptr<FdmMesher>& mesher,
+        Real sigma, Real kappa, Real theta)
+    : dyMap_(SecondDerivativeOp(1, mesher)
+                   .mult(sigma*sigma*mesher->locations(1))
+                   .add(FirstDerivativeOp(1, mesher)
+                   .mult(kappa*(theta - mesher->locations(1))))),
+      mapT_(1, mesher),
+      mesher_(mesher){
+    }
+
+    void FdmCIRRatesPart::setTime(Time t1, Time t2) {
+        mapT_.axpyb(Array(), dyMap_, dyMap_, -0.5*mesher_->locations(1));
+    }
+
+    const TripleBandLinearOp& FdmCIRRatesPart::getMap() const {
+        return mapT_;
+    }
+
+    FdmCIRMixedPart::FdmCIRMixedPart(
+        const ext::shared_ptr<FdmMesher>& mesher,
+        const ext::shared_ptr<CoxIngersollRossProcess> & cirProcess,
+        const ext::shared_ptr<GeneralizedBlackScholesProcess> & bsProcess,
+        const Real rho,
+        const Real strike)
+        : dyMap_(SecondOrderMixedDerivativeOp(0, 1, mesher)
+                     .mult(Array(mesher->layout()->size(), 2*rho*cirProcess->volatility()))),
+          mapT_(0, 1, mesher),
+          mesher_(mesher),
+          sigma1_(bsProcess->blackVolatility().currentLink()),
+          strike_(strike){
+    }
+
+    void FdmCIRMixedPart::setTime(Time t1, Time t2) {
+        const Real v = std::sqrt(sigma1_->blackForwardVariance(t1, t2, strike_)/(t2-t1));
+        const NinePointLinearOp op(dyMap_.mult(Array(mesher_->layout()->size(), v)));//TODO fix this
+        mapT_.reset(op);
```

Comment:
Since you're about to throw `op` away at the end of this method, you can write `mapT_.swap(op)`.  This removes the need for the `reset` method.

Reply:
Thanks. I'm now getting a n error

unknown location(0): fatal error: in "QuantLib test suite/Finite Difference CIR tests/QuantLib__detail__quantlib_test_case(&FdCIRTest__testFdmCIRConvergence)": std::runtime_error: Boost assertion failed: buckets_

How do I fix this? I don't get the same error when I tried running locally.